src/DvbData.cpp

```diff
@@ -590,66 +580,95 @@ const std::string &Dvb::GetLiveStreamURL(const PVR_CHANNEL &channelinfo)
   return m_channels[channelinfo.iUniqueId - 1]->streamURL;
 }
 
-
 void *Dvb::Process()
 {
   XBMC->Log(LOG_DEBUG, "%s: Running...", __FUNCTION__);
   int update = 0;
   int interval = (!g_lowPerformance) ? 60 : 300;
 
+  // set PVR_CONNECTION_STATE_CONNECTING only once!
+  SetConnectionState(PVR_CONNECTION_STATE_CONNECTING);
+
   while (!IsStopped())
   {
-    Sleep(1000);
-    ++update;
-
-    CLockObject lock(m_mutex);
-    if (m_updateEPG)
+    if (!IsConnected())
     {
-      m_updateEPG = false;
-      m_mutex.Unlock();
-      Sleep(8000); /* Sleep enough time to let the recording service grab the EPG data */
-      m_mutex.Lock();
-      XBMC->Log(LOG_INFO, "Performing forced EPG update!");
-      PVR->TriggerEpgUpdate(m_currentChannel);
-    }
+      XBMC->Log(LOG_INFO, "Trying to connect to the backend service...");
+
+      if (CheckBackendVersion() && UpdateBackendStatus(true) && LoadChannels())
+      {
+        XBMC->Log(LOG_INFO, "Connection to the backend service successful.");
+        SetConnectionState(PVR_CONNECTION_STATE_CONNECTED);
 
-    if (m_updateTimers)
+        TimerUpdates();
+        // force recording sync as Kodi won't update recordings on PVR restart
+        PVR->TriggerRecordingUpdate();
+      }
+      else
+      {
+        XBMC->Log(LOG_INFO, "Connection to the backend service failed."
+          " Retrying in 10 seconds...");
+        Sleep(10000);
+      }
+    }
+    else
     {
-      m_updateTimers = false;
-      m_mutex.Unlock();
       Sleep(1000);
```

Comment:
Please revisit the mutex locking. It always includes this Sleep which does not sound like a very good idea, does it?